runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java

```diff
@@ -57,30 +72,61 @@
 		ChatMessageType.GAME
 	);
 
+	private static final int CYCLE_HOTKEY = KeyEvent.VK_TAB;
+	private static final ImmutableSet<ChatMessageType> CYCLE_TYPES = ImmutableSet.of(
+		ChatMessageType.PRIVATE_MESSAGE_SENT,
+		ChatMessageType.PRIVATE_MESSAGE_RECEIVED,
+		ChatMessageType.PRIVATE_MESSAGE_RECEIVED_MOD
+	);
+
 	private Queue<QueuedMessage> messageQueue;
+	private Deque<String> friends;
+
+	@Inject
+	private Client client;
+
+	@Inject
+	private ClientThread clientThread;
+
+	@Inject
+	private ChatHistoryConfig config;
+
+	@Inject
+	private KeyManager keyManager;
 
 	@Inject
 	private ChatMessageManager chatMessageManager;
+
+	@Provides
+	ChatHistoryConfig getConfig(ConfigManager configManager)
+	{
+		return configManager.getConfig(ChatHistoryConfig.class);
+	}
 	
 	@Override
 	protected void startUp()
 	{
 		messageQueue = EvictingQueue.create(100);
+		friends = new ArrayDeque<>(5);
+		keyManager.registerKeyListener(this);
 	}
 
 	@Override
 	protected void shutDown()
 	{
 		messageQueue.clear();
 		messageQueue = null;
+		friends.clear();
+		friends = null;
+		keyManager.unregisterKeyListener(this);
 	}
 
 	@Subscribe
 	public void onChatMessage(ChatMessage chatMessage)
 	{
 		// Start sending old messages right after the welcome message, as that is most reliable source
 		// of information that chat history was reset
-		if (chatMessage.getMessage().equals(WELCOME_MESSAGE))
+		if (config.retainChatHistory() && chatMessage.getMessage().equals(WELCOME_MESSAGE))
```

Comment:
This is a behavior change and makes it start retaining the welcome message. Maybe don't do this.